### DISCUSSION TITLE
Submit-Button muss zweimal geklickt werden, damit Daten gerendert werden

### DISCUSSION THREAD

**Discussion by alinawinter** [neuefische/questions#547] [Web] [answered]
Hallo,

gerade arbeite ich an meiner Form Komponente mit u.a. Submit-Button, der als eigene Komponente importiert wird. Bei Drücken auf den Submit-Button sollen Daten („ResultContainer“) unter dem Formular ausgegeben werden. Das funktioniert auch, allerdings erst bei zwei mal Klicken. Es scheint also etwas mit der Einbindung des SubmitButtons in das Form oder mit der handle-Funktion nicht zu stimmen.

Ich habe meinen aktuellen Stand committed und gepusht. [Hier der Link](https://github.com/alinawinter/capstone-project/pull/5/files).

**Beschreibung des Fehlers: 
Wie kann man den Fehler reproduzieren?**
 Wenn man den Development Server startet und dann auf den Button „Info zum Tagesbedarf“ klickt, merkt man, dass man zwei mal klicken muss, um die Daten unten angezeigt zu bekommen 
**Wo und wann tritt er auf?**
 Im Browser, beim Klicken des SubmitButtons „Info zum Tagesbedarf“ 
**Was wird erwartet, was passiert?** 
Ich erwarte, dass mein SubmitButton beim ersten Klick bereits die Daten („ResultContainer“) unterhalb der Form rendert 
**Was sind die Fehlermeldungen?**
Keine

**Was habe ich bereits ausprobiert:** 
Ich habe bereits nach diesem Fehler online gesucht und habe einzelne Dinge angepasst, jetzt nach mehreren Stunden rumprobieren dachte ich, ich frage lieber mal nach.

Vielen Dank für eure Hilfe!
Alina

**Comment by PReichetanz** [comment, marked as answer]
Hallo Alina,

das Problem besteht in den Zeilen 17-20: 

```js
setSelectedFood(data.foodAte);
const selectedFoodCategory = foodCategories.find(
  (foodCategory) => foodCategory.name === selectedFood
);
```

Das Updaten des States ist mit `setSelectedFood(data.foodAte)` nur vorgemerkt, aber `selectedFood` ist immer noch ein leerer String, bis die `handleFoodSelection` in Zeile 30 fertig ist [(Caveats in React.dev docs)](https://react.dev/reference/react/useState#setstate-caveats).

Um das zu umgehen, nimmst Du einfach `data.foodAte` für den Vergleich:
```js
setSelectedFood(data.foodAte);
const selectedFoodCategory = foodCategories.find(
  (foodCategory) => foodCategory.name === data.foodAte
);
```
Damit wird der State `selectedFood` aber gar nicht mehr verwendet, also solltest Du überlegen, ob Du ihn nicht auch ganz entfernen kannst.

**Reply by alinawinter** [reply]
Juhu, das hat es gelöst. VIELEN DANK! Auch für die weitergehende Erklärung!
